src/common/utils.lua

```diff
@@ -241,48 +233,152 @@ function utils.camelCase(str)
 	return str
 end
 
-utils.notices = {}
+function utils.indexOf(t, value)
+	for i, v in ipairs(t) do
+		if v == value then
+			return i
+		end
+	end
+	return -1
+end
 
-function utils.notices.credit(msg)
-	local notice = {
-		Target = msg.From,
-		Action = "Credit-Notice",
-		Recipient = msg.Recipient,
-		Quantity = tostring(1),
+--- func description
+-- @param {table} params
+-- @param {string} params.id
+-- @param {number} params.timestamp
+-- @param {string | nil} params.owner
+-- @param {table | nil} params.controllers
+function utils.register(params)
+	assert(type(params) == "table", "Register Params must be a table")
+	local id, timestamp, owner, controllers = params.id, params.timestamp, params.owner, params.controllers
+	assert(type(id) == "string", "ANT ID must be a string")
+	assert(type(timestamp) == "number", "Timestamp must be a number")
+	assert(type(owner) == "string" or owner == nil, "Owner must be a string")
+	assert(type(controllers) == "table" or controllers == nil, "Controllers must be a table")
+	ANTS[id] = {
+		Owner = owner,
+		Controllers = controllers or {},
+		RegisteredAt = timestamp,
 	}
-	for tagName, tagValue in pairs(msg) do
-		-- Tags beginning with "X-" are forwarded
-		if string.sub(tagName, 1, 2) == "X-" then
-			notice[tagName] = tagValue
+end
+
+function utils.controllerTableFromArray(t)
+	assert(type(t) == "table", "argument needs to be a table")
+	local map = {}
+	for _, v in ipairs(t) do
+		map[v] = true
+	end
+	return map
+end
+
+function utils.updateAssociations(antId, state)
+	-- Remove previous associations for old owner and controllers
+	local oldAnt = ANTS[antId]
+
+	local newOwner = state.Owner or "nilOwner"
```

Comment:
explain this place holder, why not nil?

Reply:
a nil owner, to me, would be an invariant that we might not want to allow?

Reply:
agreed - i may need to huddle as to what is the intent here

Reply:
For the use case of undername leasing on ANT's there may be implementations that involve renouncing ownership of the ant - that said I think its prudent to allow it so long as there are also controllers on the ant. If there are neither its essentially bricked and shouldnt be tracked (no point anyways since it would never be returned in a acl call)

Reply:
in that case i would use `nil`